patches/api/0387-Paper-Plugins.patch

```diff
@@ -0,0 +1,963 @@
++            config.getDefaultPermission(),
++            Set.of(), // Aware api
++            config.getApiVersion(),
++            List.of() // Libraries
++        );
++
++        File dataFolder = new File(Bukkit.getPluginsFolder(), pluginDescriptionFile.getName());
++
++        javaPlugin.init(pluginLoader, Bukkit.getServer(), pluginDescriptionFile, dataFolder, this.pluginProvider.getSource().resolve(pluginDescriptionFile.getDisplayName()).toFile(), this, config);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/service/classloader/PaperPluginClassLoaderStorage.java b/src/main/java/io/papermc/paper/plugin/provider/service/classloader/PaperPluginClassLoaderStorage.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..672d06cb5e3c4cfd584d0c868b24891aa36480ac
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/service/classloader/PaperPluginClassLoaderStorage.java
+@@ -0,0 +1,107 @@
++package io.papermc.paper.plugin.provider.service.classloader;
++
++import io.papermc.paper.plugin.provider.configuration.PluginConfiguration;
++import org.bukkit.Bukkit;
++import org.bukkit.plugin.SimplePluginManager;
++import org.bukkit.plugin.java.LibraryLoader;
++import org.bukkit.plugin.java.PluginClassLoader;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.HashMap;
++import java.util.List;
++import java.util.Map;
++import java.util.concurrent.CopyOnWriteArrayList;
++import java.util.concurrent.locks.ReentrantReadWriteLock;
++import java.util.logging.Logger;
++
++/**
++ * Holds classloaders that other plugin providers might be using.
++ */
++public final class PaperPluginClassLoaderStorage {
++
++    private static final Logger LOGGER = java.util.logging.Logger.getLogger("PaperClassLoader");
++
++    private final Map<String, ReentrantReadWriteLock> classLoadLock = new HashMap<>();
++    private final Map<String, Integer> classLoadLockCount = new HashMap<>();
++    public final List<PaperPluginClassLoader> classLoaders = new CopyOnWriteArrayList<>();
++    public final List<PluginClassLoader> spigotClassLoaders = new CopyOnWriteArrayList<>();
++    private final LibraryLoader libraryLoader;
++
++    public static final PaperPluginClassLoaderStorage INSTANCE = new PaperPluginClassLoaderStorage();
++
++    public PaperPluginClassLoaderStorage() {
++
++        LibraryLoader libraryLoader = null;
++        try {
++            libraryLoader = new LibraryLoader(LOGGER);
++        } catch (NoClassDefFoundError ex) {
++            // Provided depends were not added back
++            LOGGER.warning("Could not initialize LibraryLoader (missing dependencies?)");
++        }
++        this.libraryLoader = libraryLoader;
++    }
++
++    public Class<?> getClassByName(String name, boolean resolve, PluginConfiguration pluginConfiguration, @Nullable ClassLoader requester) {
++        // make MT safe
++        java.util.concurrent.locks.ReentrantReadWriteLock lock;
```

Comment:
final can help here, to know that this var will be always init

Reply:
i'll deal with style stuff later